File: Algorithms/098. Validate Binary Search Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isValidBST(TreeNode root) {
    	if (root==null) {
			return true;
		}
    	List<Integer> list=inorder(root,new ArrayList<Integer>());
    	if (list.size()<2) {
			return true;
		}
    	for (int i = 0; i < list.size()-1; i++) {
			if (list.get(i)>=list.get(i+1)) {
				return false;
			}
		}
    	return true;
    }
    public List<Integer> inorder(TreeNode root,List<Integer> list){  
        if (root!=null) {  
            list=inorder(root.left,list);  
            list.add(root.val);  
            list=inorder(root.right,list);  
        }  
        return list;  
    }  
}

File: Algorithms/098. Validate Binary Search Tree/Solution.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_valid_bst(root)
    return is_valid(root,nil,nil)
end

def is_valid(root, min, max)
    return !root || ((!max || root.val< max) && (! min || root.val >min) && is_valid(root.left,min,root.val) && is_valid(root.right,root.val,max))
end